### PR TITLE
Boundary Layer Proxy Feature: wall-normal spacing signal

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Boundary layer state (attached vs separated) is the dominant factor in surface pressure distribution error, especially for p_tan (tandem transfer) and p_re (OOD Reynolds). The current model must infer boundary layer state entirely from the DSDF features and Reynolds number — it has no direct signal about the local wall-normal mesh spacing, which is a strong proxy for boundary layer resolution in CFD meshes.

**Physics motivation:** CFD mesh generators place nodes closer together in regions where the boundary layer is expected to be thin (high gradients). The wall-normal spacing to the nearest surface node directly encodes the mesh generator's expectation about boundary layer thickness. This information is available in the mesh geometry but is NOT currently used as an input feature.

**Why input features work:** Our strongest improvements have come from physics-informed input features: wake deficit (-4.1% p_in), TE coordinate frame (-5.4% p_in), gap-stagger spatial bias (-3.0% p_tan). Panel-method Cp feature showed the first p_tan improvement (-3.4%). Adding features that encode boundary layer physics follows this proven pattern.

**Hypothesis:** Two new scalar features per node will improve OOD generalization:
1. **Wall-normal distance ratio**: For each non-surface node, the ratio of its distance to the nearest surface node divided by a reference length (chord length or local mesh scale). This encodes how deep into the boundary layer the node sits.
2. **Surface proximity indicator**: A soft indicator (exponential decay) of proximity to the surface: exp(-dist/sigma) where sigma is scaled by local mesh density. For surface nodes, this is 1.0.

### Step 1: Add argument

```python
parser.add_argument('--bl_proxy_feature', action='store_true',
                    help='Add boundary layer proxy features (wall-normal distance, surface proximity)')
```

### Step 2: Compute boundary layer proxy features

This should be computed once per sample (it only depends on geometry, not on flow solution). Compute during data loading or as a cached preprocessing step.

```python
def compute_bl_proxy_features(pos, surf_mask, chord_length=1.0):
    """Compute boundary layer proxy features for each node.
    
    Args:
        pos: [N, 2] node positions
        surf_mask: [N] boolean, True for surface nodes
        chord_length: reference length for normalization
    
    Returns:
        bl_features: [N, 2] — (wall_normal_dist_ratio, surface_proximity)
    """
    surf_pos = pos[surf_mask]  # [N_surf, 2]
    
    # Distance from each node to nearest surface node
    dists = torch.cdist(pos.unsqueeze(0), surf_pos.unsqueeze(0)).squeeze(0)  # [N, N_surf]
    min_dist, _ = dists.min(dim=-1)  # [N] — distance to nearest surface node
    
    # Feature 1: wall-normal distance ratio (normalized by chord)
    wall_dist_ratio = min_dist / (chord_length + 1e-8)
    
    # Feature 2: surface proximity (exponential decay)
    # sigma = median of surface node spacing (adaptive to mesh density)
    surf_spacing = torch.cdist(surf_pos.unsqueeze(0), surf_pos.unsqueeze(0)).squeeze(0)
    surf_spacing.fill_diagonal_(float('inf'))
    median_spacing = surf_spacing.min(dim=-1).values.median()
    surface_proximity = torch.exp(-min_dist / (2 * median_spacing + 1e-8))
    
    return torch.stack([wall_dist_ratio, surface_proximity], dim=-1)  # [N, 2]
```

### Step 3: Integrate as input features

Concatenate the 2 new features to the input tensor:

```python
if args.bl_proxy_feature:
    # Compute for each sample in the batch
    bl_feats = compute_bl_proxy_features(
        pos=x[:, :, :2],  # Node positions (first 2 channels)
        surf_mask=surf_mask,
        chord_length=1.0  # Normalize; adjust if chord length varies
    )
    x = torch.cat([x, bl_feats], dim=-1)  # Append as new input channels
```

**Performance note:** `torch.cdist` on [N, N_surf] may be slow for large N. If N > 5000, consider:
- Computing BL features once during data preprocessing and caching them
- Using a KD-tree or approximate nearest neighbor instead of brute-force cdist
- Only computing for the first 2000 nearest surface nodes

### Step 4: Run 2 seeds

```bash
cd cfd_tandemfoil && python train.py \
  --agent alphonse --wandb_name "alphonse/bl-proxy-s42" --seed 42 \
  --wandb_group boundary-layer-proxy \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling \
  --bl_proxy_feature

# Repeat for seed 73: --seed 73 --wandb_name "alphonse/bl-proxy-s73"
```

**Watch for:** If torch.cdist causes OOM or slow epochs (>90s/epoch), implement the caching approach. Current baseline is ~75s/epoch.

## Baseline

| Metric | Baseline | Target to beat |
|--------|----------|----------------|
| p_in | 11.742 | < 11.742 |
| p_oodc | 7.643 | < 7.643 |
| p_tan | 27.874 | < 27.874 |
| p_re | 6.419 | < 6.419 |

- **Baseline W&B runs:** k5qwvce4 (seed 42), 7oa5xfhi (seed 73)
- **val/loss baseline:** ~0.37

### CODE REVIEW

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2330 because ADVISOR: Closing — BL proxy feature regresses all metrics severely (p_in +23.5%, p_oodc +18.9%, p_tan +5.7%, p_re +15.0%).

The wall-normal distance and surface proximity features likely introduced noise or computational overhead that degraded training. The torch.cdist computation for all nodes × all surface nodes may have slowed training significantly (fewer effective epochs), and the resulting features may be too noisy on our non-uniform meshes to be useful.

**Root cause:** The DSDF features already encode distance-from-surface information (that's literally what signed distance functions provide). The BL proxy features are largely redundant with DSDF but computed less accurately. The model doesn't need two different encodings of the same geometric property.

Moving you to a fresh experiment..

**Comment by tcapelle** [comment]
STUDENT alphonse:

## Results

### Boundary Layer Proxy Feature — 2-Seed Run

| Metric | Seed 42 | Seed 73 | 2-seed avg | Baseline | Δ |
|--------|---------|---------|------------|----------|---|
| **p_in** | 15.1 | 13.9 | **14.500** | 11.742 | **+23.5%** ❌ |
| **p_oodc** | 8.9 | 9.3 | **9.100** | 7.643 | **+19.1%** ❌ |
| **p_tan** | 29.5 | 29.5 | **29.500** | 27.874 | **+5.8%** ❌ |
| **p_re** | 7.3 | 7.5 | **7.400** | 6.419 | **+15.3%** ❌ |
| **val/loss** | 0.4125 | 0.4094 | **0.4110** | ~0.37 | **+11.1%** ❌ |
| **Best epoch** | 121 | 122 | — | ~149 | — |

**W&B runs:** [ve85sqfu](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/ve85sqfu) (s42), [6na7zafa](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/6na7zafa) (s73)

**Command:**
```bash
cd cfd_tandemfoil && python train.py --agent alphonse --seed 42 \
  --wandb_name "alphonse/bl-proxy-s42" --wandb_group "boundary-layer-proxy" \
  --bl_proxy_feature \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep --residual_prediction \
  --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias --dct_freq_loss --dct_freq_weight 0.05 \
  --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling
```

### What happened

**All metrics regressed** — p_in (+23.5%), p_oodc (+19.1%), p_tan (+5.8%), p_re (+15.3%).

**Critical confound: training only reached 121-122 epochs (vs ~149 baseline).**

The BL proxy feature computation (`torch.cdist` per sample per batch, plus surface-to-surface NN distances) added ~20% per-epoch overhead. With the 180-minute wall-clock timeout:
- Baseline: ~149 epochs → cosine LR schedule completes at T_max=150, EMA eval model created at epoch 140
- BL proxy: ~121 epochs → LR schedule never completes, EMA eval model never created

This means the comparison is severely confounded — the model never reached its peak performance phase. The same issue occurred with the EMA Teacher experiment (PR #2318).

**Physics analysis:**

The two features — log1p(wall_dist_ratio/chord) and exp(-dist/(2*median_spacing)) — are partially redundant with the existing `dsdf(8)` features which already encode distances to domain boundaries, and with the `dist_feat = log1p(dsdf_min * 10)` feature already in the pipeline. The BL proxy adds nearest-surface-node distance as a more exact pointwise measure, but the signal content overlaps heavily with what the model already has.

The surface proximity feature (exponential decay) decays to <0.01 within ~5 cell spacings from the surface, making it effectively a binary near-wall indicator rather than a smooth BL-thickness proxy for most volume nodes.

**Minor note:** Post-training verification section produced NaN values (debug runs showed "Match: YES" — likely a pre-existing verification bug unrelated to BL proxy, as it also happened in #2323 without these features).

### Suggested follow-ups

1. **Precompute BL features at dataset load time** — eliminates per-epoch overhead, enabling a fair comparison. The features are deterministic functions of geometry only (pos + is_surface). Could cache them alongside the preprocessed samples.

2. **If the overhead can't be eliminated, adjust cosine_T_max** — set T_max=120 to ensure the LR schedule completes within the reduced epoch budget. Or alternatively, reduce the model to n_layers=2 to free up per-epoch budget.

3. **Consider dropping this feature entirely** — the heavy redundancy with existing dsdf + dist_feat features suggests limited new information. The most productive direction for OOD improvements has been physics-informed geometric features (TE coord frame, wake deficit) rather than generic mesh-proximity features.
